server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessor.java

```diff
@@ -225,14 +225,14 @@ static ForwardDecision removeRecipient(MailAddress recipient) {
 
         static ForwardDecision sendACopy(MailetContext context,
                                          MailAddress originalRecipient,
-                                         LoopPrevention.RecordedRecipients recordedRecipients,
                                          Set<MailAddress> newRecipients) {
             return mail -> {
                 MailImpl copy = MailImpl.duplicate(mail);
+                LoopPrevention.RecordedRecipients recordedRecipients = LoopPrevention.RecordedRecipients.fromMail(mail);
                 try {
                     copy.setSender(originalRecipient);
                     copy.setRecipients(newRecipients);
-                    recordedRecipients.mergeIfEmpty(originalRecipient).merge(newRecipients).recordOn(copy);
+                    recordedRecipients.merge(originalRecipient).recordOn(copy);
```

Comment:
Setup: A forward to B, A to C, B to C, C to B
if we only save originalRecipient to the list, we could not prevent B and C from resending email to each other. However, if you want to make it simple, saving only originalRecipient is still acceptable.

Reply:
With the new logic you just need to carry over the original recipient. Easier...